### PR TITLE
CLOUDP-139997: Add the command atlas cluster advancedSettings update to AtlasCLI

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-139997

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

```
./bin/atlas cluster advancedSettings update Cluster69449 --oplogSizeMB 1222                                                                                                1.18
Updating advanced configuration settings of your cluster'.
```

```
./bin/atlas cluster advancedSettings update --help                                                                                                                         1.18
Update advanced configuration settings for one cluster.

Usage:
  atlas clusters advancedSettings update <clusterName> [flags]

Examples:
  Update oplog minimum size for a cluster:
  $ atlas cluster advancedSettings update <clusterName> --projectId <projectId> --oplogSizeMB 1000

  Update minimum TLS protocol for a cluster:
  $ atlas cluster advancedSettings update <clusterName> --projectId <projectId> --minimumEnabledTLSProtocol "TLS1_0"

Flags:
      --disableFailIndexKeyTooLong             Flag that indicates whether you can insert or update documents where all indexed entries don't exceed 1024 bytes. If you provide this flag, mongod writes documents that exceed this limit but doesn't index them.
      --disableJavascript                      Flag that indicates whether the cluster allows execution of operations that perform server-side executions of JavaScript.
  -h, --help                                   help for update
      --noTableScan                            Flag that indicates whether the cluster disables executing any query that requires a collection scan to return results.
      --oplogMinRetentionHours float           Minimum retention window for cluster's oplog expressed in hours.
      --oplogSizeMB int                        Storage limit of cluster's oplog expressed in megabytes.
  -o, --output string                          Output format. Valid values are json, json-path, go-template, or go-template-file.
      --projectId string                       Project ID to use. Overrides the settings in the configuration file or environment variable.
      --readConcern string                     Default level of acknowledgment requested from MongoDB for read operations set for this cluster. (default "available")
      --sampleRefreshIntervalBIConnector int   Interval in seconds at which the mongosqld process re-samples data to create its relational schema.
      --sampleSizeBIConnector int              Number of documents per database to sample when gathering schema information. (default 1000)
      --tlsProtocol string                     Minimum Transport Layer Security (TLS) version that the cluster accepts for incoming connections.
      --writeConcern string                    Default level of acknowledgment requested from MongoDB for write operations set for this cluster. (default "1")

Global Flags:
  -P, --profile string   Profile to use from your configuration file.
```

Waiting for https://github.com/mongodb/go-client-mongodb-atlas/pull/316

### CODE REVIEW

File: internal/cli/atlas/clusters/advancedsettings/update.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+func (opts *UpdateOpts) newProcessArgs() *atlas.ProcessArgs {
+	args := &atlas.ProcessArgs{
+		DefaultReadConcern:               opts.defaultReadConcern,
+		DefaultWriteConcern:              opts.defaultWriteConcern,
+		MinimumEnabledTLSProtocol:        opts.minimumEnabledTLSProtocol,
+		SampleSizeBIConnector:            pointy.Int64(opts.sampleSizeBIConnector),
+		SampleRefreshIntervalBIConnector: pointy.Int64(opts.sampleRefreshIntervalBIConnector),
+		NoTableScan:                      pointy.Bool(opts.noTableScan),
+	}
```
Comment (gssbzn) [inline internal/cli/atlas/clusters/advancedsettings/update.go:82]: is this a PATCH or a PUT? you may be over-writing settings here by defining all the options

Comment (gssbzn) [reply internal/cli/atlas/clusters/advancedsettings/update.go:82]: to be clear what happens if I do
```bash
atlas cluster advancedSettings update <clusterName> --projectId <projectId> --writeConcern majority

atlas cluster advancedSettings update <clusterName> --projectId <projectId> --readConcern  majority
```

what's the value of `writeConcern` after the second call?

Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go:82]: it is a [PATCH](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#operation/updateAdvancedConfigurationOptionsForOneCluster)

Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go:82]: You are right:
```
./bin/atlas cluster advancedSettings update Cluster87819 --writeConcern majority

~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                            1.18
{
  "defaultReadConcern": "available",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}

./bin/atlas cluster advancedSettings update Cluster87819 --readConcern  local                                                                                                 1.18
Updating advanced configuration settings of your cluster'.


./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                            1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "1",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
```

Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go]: Updated. Now it is working as expected.

```
./bin/atlas cluster advancedSettings update Cluster87819 --writeConcern majority                                                                                           1.18
Updating advanced configuration settings of your cluster'.
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                         1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings update Cluster87819 --readConcern  local                                                                                              1.18
Updating advanced configuration settings of your cluster'.
~/workspace/mongodb-atlas-cli  CLOUDP-139997 *1 +1 ❯ ./bin/atlas cluster advancedSettings describe Cluster87819 -o json                                                                                                         1.18
{
  "defaultReadConcern": "local",
  "defaultWriteConcern": "majority",
  "minimumEnabledTlsProtocol": "TLS1_2",
  "javascriptEnabled": true,
  "noTableScan": false,
  "sampleSizeBIConnector": 1000,
  "sampleRefreshIntervalBIConnector": 0
}
```

**Review by gssbzn — COMMENTED** [review]
for all the disable there should be an enable, same with the noTableScan to allow people to enable them again

**Comment by gssbzn** [comment]
> for all the disable there should be an enable, same with the noTableScan to allow people to enable them again

also remember to make them mutually exclusive

**Review by gssbzn — DISMISSED** [review]
LGTM thanks for all the work here

File: internal/cli/atlas/clusters/advancedsettings/update.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+func UpdateBuilder() *cobra.Command {
+	opts := &UpdateOpts{}
+	cmd := &cobra.Command{
+		Use:   "update <clusterName>",
+		Short: "Update advanced configuration settings for one cluster.",
+		Example: fmt.Sprintf(`  Update oplog minimum size for a cluster:
+  $ %[1]s cluster advancedSettings update <clusterName> --projectId <projectId> --oplogSizeMB 1000
+
+  Update minimum TLS protocol for a cluster:
+  $ %[1]s cluster advancedSettings update <clusterName> --projectId <projectId> --minimumEnabledTLSProtocol "TLS1_0"`,
```
Comment (jwilliams-mongo) [inline internal/cli/atlas/clusters/advancedsettings/update.go:148]: suggestion: I think TLS 1.0 is deprecated and known to be insecure. consider using a different version in the example 
```suggestion
  $ %[1]s cluster advancedSettings update <clusterName> --projectId <projectId> --minimumEnabledTLSProtocol "TLS1_2"`,
```

**Review by jwilliams-mongo — APPROVED** [review]
@andreaangiolillo LGTM with a few minor suggestions

File: internal/cli/atlas/clusters/advancedsettings/update.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+// atlas cluster(s) advancedSettings update <clusterName> --projectId projectId [--readConcern readConcern]
+// [--writeConcern writeConcern] [--disableFailIndexKeyTooLong true/fale] [--enableFailIndexKeyTooLong true/fale] [--disableJavascript true/false] [--enableJavascript true/false]
+// [--minimumEnabledTLSProtocol minimumEnabledTLSProtocol] [--disableTableScan true/false] [--enableTableScan true/false] [--oplogMinRetentionHours oplogMinRetentionHours]
+// [--oplogSizeMB oplogSizeMB] [--sampleRefreshIntervalBIConnector sampleRefreshIntervalBIConnector] [--sampleSizeBIConnector sampleSizeBIConnector].
+func UpdateBuilder() *cobra.Command {
+	opts := &UpdateOpts{}
+	cmd := &cobra.Command{
+		Use:   "update <clusterName>",
+		Short: "Update advanced configuration settings for one cluster.",
+		Example: fmt.Sprintf(`  Update oplog minimum size for a cluster:
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go]: updated

File: internal/cli/atlas/clusters/advancedsettings/update.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  $ %[1]s cluster advancedSettings update <clusterName> --projectId <projectId> --oplogSizeMB 1000
+
+  Update minimum TLS protocol for a cluster:
```
[the remark this answers is not in the document]
Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go]: updated

File: internal/cli/atlas/clusters/advancedsettings/update.go (hunk 2 above)
Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go]: updated

File: internal/cli/atlas/clusters/advancedsettings/update.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	"github.com/openlyinc/pointy"
+	"github.com/spf13/cobra"
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	updateTmpl                   = "Updating advanced configuration settings of your cluster'.\n"
+	defaultReadConcern           = "available"
+	defaultWriteConcern          = "1"
+	defaultSampleSizeBIConnector = 1000
```
Comment (gssbzn) [inline internal/cli/atlas/clusters/advancedsettings/update.go:37]: I'm catching just now that I think these should be removed, you are doing and if this value then don't set it for the request, which means if I change the read concern and want to go back to the default I can't, right?

Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go:37]: yes, good catch!

Comment (andreaangiolillo) [reply internal/cli/atlas/clusters/advancedsettings/update.go]: updated

**Review by gssbzn — APPROVED** [review]
LGTM

**Pull request merged**